chirp/drivers/uvk5.py

```diff
@@ -0,0 +1,1119 @@
+              612, 624, 627, 631, 632, 654, 662, 664, 703, 712, 723,
+              731, 732, 734, 743, 754
+              ]
+
+FLOCK_LIST=[ "Off", "FCC", "CE", "GB", "430", "438" ]
+SCANRESUME_LIST = [ "TO: Resume after 5 seconds", "CO: Resume after signal dissapears", "SE: Stop scanning after receiving a signal" ]
+WELCOME_LIST = [ "Full Screen", "Welcome Info", "Voltage" ]
+KEYPADTONE_LIST = [ "Off", "Chinese", "English" ]
+LANGUAGE_LIST = [ "Chinese", "English" ]
+ALARMMODE_LIST = [ "SITE", "TONE" ]
+REMENDOFTALK_LIST = [ "Off" , "ROGER", "MDC" ]
+RTE_LIST = [ "Off" , "100ms", "200ms", "300ms", "400ms", "500ms", "600ms", "700ms", "800ms", "900ms" ]
+
+MEM_SIZE=0x2000 #size of all memory
+PROG_SIZE=0x1d00 #size of the memowy that we will program
+MEM_BLOCK=0x80 #largest block of memory that we can reliably write
+
+# the communication is obfuscated using this fine mechanism
+def xorarr(data: bytes):
+    tbl=[ 22, 108, 20, 230, 46, 145, 13, 64, 33, 53, 213, 64, 19, 3, 233, 128]
+    #x=[]
+    x=b""
+    r=0
+    for byte in data:
+        x+=bytes([byte^tbl[r]])
+        r=(r+1)%len(tbl)
+    return x
+
+# if this crc was used for communication to AND from the radio, then it
+# would be a measure to increase reliability.
+# but it's only used towards the radio, so it's for further obfuscation
+def calculate_crc16_xmodem(data: bytes):
+    poly=0x1021
+    crc = 0x0
+    for byte in data:
+        crc = crc ^ (byte << 8)
+        for i in range(8):
+            crc = crc << 1
+            if (crc & 0x10000):
+                crc = (crc ^ poly ) & 0xFFFF
+    return crc & 0xFFFF
+
+
+def _send_command(radio,data: bytes):
+    """Send a command to UV-K5 radio"""
+    serial=radio.pipe
+    serial.timeout=0.5
+
+    crc=calculate_crc16_xmodem(data)
+    data2=data+bytes([crc&0xff,(crc>>8)&0xff])
+
+    command=b"\xAB\xCD"+bytes([len(data)])+b"\x00"+xorarr(data2)+b"\xDC\xBA"
+#    print("Sending command: %s" % util.hexprint(command))
+    serial.write(command)
+
+def _receive_reply(radio):
+    serial=radio.pipe
+    serial.timeout=0.5
+    headed: bytes
+    cmd: bytes
+    footer: bytes
+
+    header=serial.read(4)
+    if header[0]!=0xAB or header[1]!=0xCD or header[3]!=0x00:
+        if PRINT_CONSOLE:
+            print("bad response header")
+        raise errors.RadioError( "Bad response header")
+ 
+        return False
+
+    cmd=serial.read(int(header[2]))
+    footer=serial.read(4)
+
+    if footer[2]!=0xDC or footer[3]!=0xBA:
+        if PRINT_CONSOLE:
+            print("bad response footer")
+        raise errors.RadioError( "Bad response footer")
+
+        return False
+
+    cmd2=xorarr(cmd)
+    
+    if PRINT_CONSOLE:
+        print("Received reply: %s" % util.hexprint(cmd2))
+
+    return cmd2
+
+
+def _getstring(data: bytes, begin, maxlen):
+    s=""
+    c=0
+    for i in data:
+        c+=1
+        if c<begin:
+            continue
+        if i<ord(' ') or i>ord('~'):
+            break
+        s+=chr(i)
+    return s
```

Comment:
It would really be better to just slice the bytestring and then decode it to ascii or whatever you expect. This is rather unconventional.

Reply:
Changed to a pythonized version, although i think this was more readable